### PR TITLE
Save link to current recipe in bookmarks

### PR DESCRIPTION
**Pull request by slprime**
How it work
When you add an item to the bookmarks, NEI remembers the recipe tab and recipe number. So when you select an item from bookmarks, you get into that tab and to that recipe.

Related issue
#47

How it work

Save current recipe when add item to bookmarks
This behavior can be disabled in settings
If need open item without open related recipe - use shift key and click to item
Related recipe open not only from bookmarks panel
If Recipe move, remove or add new recipes in modpack - link to the recipe work correctly or open first tab and first page

### CODE REVIEW

File: src/codechicken/nei/recipe/GuiRecipe.java [hunk 1]
```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
...
 
+    public void openTargetRecipe(String recipeId) 
+    {
+
+        if (recipeId == "") {
+            return;
+        }
+
+        String[] expected = recipeId.split("\\|", 2);
+        Integer recipetype = 0;
```
Comment (mitchej123) [inline src/codechicken/nei/recipe/GuiRecipe.java:223]: why `Integer` vs `int`?

File: src/codechicken/nei/recipe/GuiRecipe.java [hunk 2]
```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
...
+        page = Math.min(idp, (handler.numRecipes() - 1) / getRecipesPerPage());
         recipeTabs.calcPageNumber();
         checkYShift();
         initOverlayButtons();
     }
 
+    public void openTargetRecipe(String recipeId) 
+    {
+
+        if (recipeId == "") {
```
Comment (mitchej123) [inline src/codechicken/nei/recipe/GuiRecipe.java:218]: `.isEmpty()`

File: src/codechicken/nei/recipe/GuiRecipe.java [hunk 3]
```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
...
+            return;
+        }
+
+        String[] expected = recipeId.split("\\|", 2);
+        Integer recipetype = 0;
+        Integer page = 0;
+
+        for (int j = 0; j < currenthandlers.size(); j++) {
+            IRecipeHandler localHandler = currenthandlers.get(j);
```
Comment (mitchej123) [inline src/codechicken/nei/recipe/GuiRecipe.java:227]: nit: final

File: src/codechicken/nei/recipe/GuiRecipe.java [hunk 4]
```diff
@@ -193,20 +195,140 @@ private void checkYShift() {
...
+            HandlerInfo localHandlerInfo = getHandlerInfo(localHandler);
+
+            if (localHandlerInfo.getHandlerName().equals(expected[0])) {
+                recipetype = j;
+
+                for (int i = 0; i < localHandler.numRecipes(); i++) {
+                    if (createRecipeId(localHandler, i).equals(expected[1])) {
```
Comment (mitchej123) [inline src/codechicken/nei/recipe/GuiRecipe.java:234]: This seems... very heavy handed on the allocation side to find the right page. We're creating a string to compare, and then throwing it away.  Is there another ID we can use that already exists, or some way to cache this?  Some of these handlers can have a large number of items and this has the potential to re-introduce slowness back into NEI.

**Comment by mitchej123** [comment]
The functionality seems reasonable, I'm a bit concerned about the performance though, especially on handlers with a lot of items (GT5, for example)

File: src/codechicken/nei/recipe/GuiRecipe.java (hunk 3 above)
Comment (slprime) [reply src/codechicken/nei/recipe/GuiRecipe.java:227]: here I do not understand what you mean. Whee need `final`? in 227 line?

File: src/codechicken/nei/recipe/GuiRecipe.java (hunk 4 above)
Comment (slprime) [reply src/codechicken/nei/recipe/GuiRecipe.java:234]: I tried to compare nbt, but the code became too complicated to understand. Also more strings will be created on handlers with more variants for recipes for one item. And this is not so many handlers

File: src/codechicken/nei/recipe/GuiRecipe.java (hunk 1 above)
Comment (slprime) [reply src/codechicken/nei/recipe/GuiRecipe.java]: fix

File: src/codechicken/nei/recipe/GuiRecipe.java (hunk 2 above)
Comment (slprime) [reply src/codechicken/nei/recipe/GuiRecipe.java]: fix

**Comment by slprime** [comment]
New PR: https://github.com/GTNewHorizons/NotEnoughItems/pull/174

**Pull request closed without merging**